### PR TITLE
Add support for processing keywords

### LINKED ISSUE #25: Add support for processing such as cutouts
This would allow people to use the cutout service etc.

See [this conversation](https://matrix.to/#/!HZAyHIffAnrJIWEjtH:matrix.org/$15156044671100488iXtVD:matrix.org?via=matrix.org&via=cadair.com) in matrix for some context.

### PR DESCRIPTION
**Pull request by wtbarnes**
Fixes #25 

This was easier than expected though I'm not totally sure about the desired API here. In theory, this will support all of the processing keywords, though I've only tested it with `im_patch`. 

This works by passing in a dictionary to `Client.export` where each key in the dict corresponds to the process keyword and the value is another dict with the appropriate arguments for that keyword. For example, for the `im_patch` option for cutouts,

```python
import drms

c = drms.Client()

tstart = '2015-10-17T04:33:30.000'
tstop = '2015-10-17T04:34:30.000'
tref = tstart
duration = '1m'
cadence = '12s'
ds = f'aia.lev1_euv_12s[{tstart}/{duration}@{cadence}][171]{{image}}'
notify = 'first.last@email.com'

im_patch = {
    't_start': tstart,
    't_stop': tstop,
    't_ref': tref,
    'cadence': cadence,
    't': 0,
    'r': 0,
    'c': 0,
    'locunits': 'arcsec',
    'boxunits': 'arcsec',
    'x': -517.2,
    'y': -246,
    'width': 345.6,
    'height': 345.6,
}

resp = c.export(
    ds,
    email=notify,
    method='url',
    protocol='fits',
    process={'im_patch': im_patch},
)
print(resp)
resp.wait()
resp.download('./')
```

### CODE REVIEW

**Comment by wtbarnes** [comment]
I suppose this probably needs some tests...

**Comment by wtbarnes** [comment]
I've also used f-strings which I realize are not compatible with the supported versions of Python, but maybe this is ok once #41 is merged.

**Comment by mbobra** [comment]
Is the goal here to support all of the processing options, or only to enable the cutouts (using `im_patch`)?

All the processing options include:
* `aia_scale` - **Scale the image to 0.6 arcsec/pixel**. Users can also extract a cutout using this option (which makes it the equivalent of using `rebin` and `im_patch` together).
* `HmiB2ptr` - **Converts the image to r, θ, φ coordinates (from azimuth, inclination, and field)**. This only applies for the drms series hmi.B_720. 
* `resize` - **Resize and rotate the image using sub-pixel registration**. Users can co-register images to the Sun center or to the first image in the query; users can also interpolate the images, change the scale, or set all pixels off the limb to NaN. Some of these options are only available for hmi.* drms series.
* `im_patch` - **Extract a partial-disk image from a full-disk image**. If the query spans multiple images, users can either track an image at the Carrington rotation rate or extract the same patch of Sun from each image. 
* `maproj` - **Extract a partial-disk image from a full-disk image and remap it to a Plate Carree, Cylindrical Equal Area, Orthographic, or Postel projection**.
* `rebin` - **Rebin an image using boxcar or Gaussian smoothing**.

Users can also select multiple options; e.g. select `im_patch` and `rebin` together.

**Comment by wtbarnes** [comment]
Yes, that's what I was thinking. `process` is a dictionary where each key is the name of the process and the value is a dictionary with the appropriate input arguments.

If a user knows what keywords and the right names of the processes, any of those, and any combination of them, should be able to be passed. 

The only validation that is done is checking the process keyword and requiring it to be one of those you've listed there.

**Comment by davisite** [comment]
@wtbarnes, @mbobra 

`jsoc_fetch` supports applying more than one processing type at the same time; for example, to both resize and rebin, the `process` argument to jsoc_fetch would look like:

process=n=0|resize,regrid=1,do_stretchmarks=0,center_to=0,rescale=1,scale_to=1|rebin,u=0,scale=0.5,method=boxcar

This first resizes the images, then rescales them. I describe a technique for figuring out the `process` value in https://github.com/sunpy/drms/issues/25

**Comment by wtbarnes** [comment]
This PR does allow for multiple processing steps per export request. The user would just add each of these steps to the dictionary passed to the `process` kwarg with the appropriate key. 

The approach you describe in #25 using exportdata is actually exactly what I did 😅.

**Comment by mbobra** [comment]
I verified the test above. I also performed another random test (below) to rebin the image, and it worked. I didn't test this PR extensively but I think the biggest problem is the documentation; when I constructed the test below, I accidentally wrote `sizeratio` instead of `scale` and the export completed, but the image was still the same scale (4096x4096).

Here is some documentation for [im_patch](http://jsoc.stanford.edu/doxygen_html/group__im__patch.html), [rebin](http://jsoc.stanford.edu/doxygen_html/group__jsoc__rebin.html), and [Maproj](http://jsoc.stanford.edu/doxygen_html/export__as__maproj_8c-source.html). I could not find documentation for resize and aia_scale, and this PR doesn't support hmib2ptr. 

Also, do we need to support aia_scale? We have other SunPy tools to scale AIA images to 0.6 arcsec/pixel. Similarly, should we support HmiB2ptr? Also happy to leave that to another PR. 

Anyway, I'm happy to approve this PR, as long as we know [1] it gives little to no error messages, and therefore an export could complete but not give you what you asked for (or fail without telling you why), and [2] we don't have sufficient documentation to guide users.

I do think it is nice to have this functionality in `drms` so thank you for all this work ☀️ 

Here's my test:
```python
tstart = '2020-10-17_22:12:00_TAI'
tref = tstart
duration = '24m'
ds = f'hmi.M_720s[{tstart}/{duration}]{{magnetogram}}'
notify = 'my@email.address'

jsoc_rebin = {
    'method': 'boxcar',
    'scale': 0.25,
}

resp = c.export(
    ds,
    email=notify,
    method='url',
    protocol='fits',
    process={'rebin': jsoc_rebin},
)

print(resp)
resp.wait()

resp.download('./')
```

**Comment by wtbarnes** [comment]
> Also, do we need to support aia_scale? We have other SunPy tools to scale AIA images to 0.6 arcsec/pixel. Similarly, should we support HmiB2ptr? Also happy to leave that to another PR.

I think that's fine. I view this package as being completely independent of sunpy. I think the aim should be to support whatever JSOC/drms supports. On the sunpy side, we can then choose to support whatever subset of these features is appropriate via `JSOCClient`. 

>  but I think the biggest problem is the documentation;

I very much agree. Do we want to use this PR as an opportunity to better document the available process keywords and the associated arguments? It would be fairly straightforward to just have a list of the allowed arguments for each process. The trickier part would be keeping this up to date with upstream changes in JSOC.

File: drms/json.py
Comment (davisite) [inline drms/json.py:403]: I see some 'missing' elements in the `allowed_processes` list in _exp_request_url(...). In addition to 'aia_scale_aialev1', there are 'aia_scale_orig' and 'aia_scale_other'. And, as @mbobra has pointed out, 'HmiB2ptr' was also omitted. There is no reason all processing types need be present, but I thought I'd mention ones that might be not well-known.

There is an algorithm used to choose between the three aia_scale variants. First, all variants operate only on certain series. SunPy should disable all variants when any of these conditions are false:

* the series' images are 2D
* keyword CTYPE1 = 'HPLN-TAN'
* keyword CTYPE2 = 'HPLN-TAN'

Second, if aia_scale processing is being performed on aia.lev1, the proper variant is 'aia_scale_orig'. If the processing is being performed on any other aia lev1 series, then the proper variant is 'aia_scale_aialev1'. For all other series where the three restrictions listed above hold, the variant is 'aia_scale_other'.

If the correct variant to use is 'aia_scale_orig', then SunPy should force the value of the filename format variable (`filenamefmt`) to 'aia.lev1.{T_OBS:A}.{WAVELNTH}.{segment}'. Otherwise, the export attempt will fail. To determine if a series is a level-1 series, there needs to be a check for the value of another keyword - KEYWORD_LVL_NUM.

Since choosing the correct variant requires performing this cumbersome algorithm, it might be desirable to simply not support any of the aia_scale processing types. OR, it might be desirable to support the aia_scale_orig variant only, and only if the series is aia.lev1. aia_scale_orig operating on aia.lev1 results in aia.lev1p5 data (level 1.5 aia data).

Likewise, 'HmiB2ptr' operates only under certain conditions, on certain series. I can provide that algorithm if it is wanted, but again, if it is not needed, then I would not bother.

**Review by davisite — COMMENTED** [review]
Hello - I was asked if I wanted to review the changes; the py code changes look good to me, but please see my general comment about the types of processing supported.

File: drms/json.py
Comment (wtbarnes) [reply drms/json.py:403]: Thanks Art! This is really helpful. For some reason, I was thinking I had stuck `HmiB2ptr` in there, but clearly I left it out on accident.

Checking the series name at this point and adjusting the the `aia_scale_*` command should be relatively straightforward. However, I'm not sure about the keyword checks or how to know which series of images are 2D. 

Given all of these complications and of course the possibility of upstream changes to how the JSOC performs these processing operations, I'm hesitant to put really restrictive checks on all processing options. Maybe we should just view this as a flexible, albeit unchecked interface to the processing keywords.

Ideally, users who want a higher level interface to these options will use sunpy (e.g. sunpy/sunpy#4595) and at that stage, we will do much more careful input validation.

Comment (wtbarnes) [reply drms/json.py:403]: I've added the additional `aia_scale_*` keywords to the allowed processing key list

**Comment by wtbarnes** [comment]
This should still probably wait on #41 because of the use of f-strings.

File: drms/json.py
Comment (davisite) [reply drms/json.py:403]: One half-formed idea I had to consider in the future (I would go ahead with what you have now) would be for the JSOC to provide the list of applicable processing types. Maybe that would get returned to the drms package by the jsoc_info AJAX call. That way the JSOC can evaluate the very-JSOC-specific conditions for the focal series and determine which processing steps to offer. But that does present the issue that the drms client would need to call one of the client API functions that in turn calls the jsoc_info AJAX. The client could not directly call export() without first doing this.

I should note that the JSOC and our export-page and export-script users always invoke jsoc_info first. It is a requirement on the JSOC export web page.

This type of solution, where the drms package can remove itself from some specialized JSOC thing, deferring to the JSOC, is going to be my general suggestion for issues like this. This was the rationale behind my suggestion that the drms package not append the [? quality >= 0 ?] filter to record-set specifications - let the JSOC determine when to add this filter.

Comment (wtbarnes) [reply drms/json.py:403]: Ah ok. Yes, that sounds like the most optimal solution and of course would reduce the burden of downstream packages like `drms` if we could use JSOC itself to validate the API. How long do you think the timeline would be for a change like that? Should we go ahead with the solution implemented here and then make the changes to use that information from `jsoc_info` at a later date?

Comment (davisite) [reply drms/json.py:403]: Go ahead with what you have now; I don't have an ETA for this - I probably will not be able to get to it for a while.

Comment (wtbarnes) [reply drms/json.py:403]: 👍 works for me.

**Comment by nabobalis** [comment]
I have rebased this on top of master now. I added a no change log entry only because I manually created a release changelog and I think you should just add to it instead of using towncrier this time.

File: CHANGELOG.rst
Comment (nabobalis) [inline CHANGELOG.rst:24]:
```suggestion
- See the "Processing" section of the `JSOC Data Export page <http://jsoc.stanford.edu/ajax/exportdata.html>`__ for more information.
```

**Comment by wtbarnes** [comment]
I've added a test and an example. I think this is ready to go.

File: CHANGELOG.rst
Comment (nabobalis) [inline CHANGELOG.rst:17]:
```suggestion
- ``Client.get()`` has been removed, use :meth:`drms.client.Client.query()` instead.
```

File: CHANGELOG.rst
Comment (nabobalis) [inline CHANGELOG.rst:22]:
```suggestion
- :meth:`drms.client.Client.Client.export()` now accepts a ``process`` keyword argument
```
This should render.

File: drms/json.py
Comment (nabobalis) [inline drms/json.py:421]:
```suggestion
                    raise ValueError(f'{p} is not one of the allowed processing options: {allowed_processes}')
```

File: drms/json.py
Comment (Cadair) [inline drms/json.py:406]: f-string

File: examples/cutout_export_request.py
Comment (Cadair) [inline examples/cutout_export_request.py:23]: pathlib?

Comment (nabobalis) [reply examples/cutout_export_request.py:23]: All the examples will need updating.

**Comment by nabobalis** [comment]
Is it possible to test all the keywords in one go?

File: examples/cutout_export_request.py
Comment (wtbarnes) [reply examples/cutout_export_request.py:23]: Yeah I would just hold off on updating all the gallery examples until later

File: drms/json.py
Comment (wtbarnes) [inline drms/json.py:406]:
```suggestion
        d['process=n'] = f'{n}'
```

**Comment by wtbarnes** [comment]
> Is it possible to test all the keywords in one go?

I don't really know how to do this. There are currently tests for `rebin` and for `im_patch`. We should obviously have tests for others, but the APIs are not well-documented and since the export requests fail silently, it is very difficult to test whether this processing step actually succeeds.

**Comment by nabobalis** [comment]
Some docbuild issues:
```
/home/vsts/work/1/s/.tox/build_docs/lib/python3.8/site-packages/drms/client.py:docstring of drms.client.Client.export:47: WARNING: py:obj reference target not found: drms
writing output... [ 50%] api/drms.client.ExportRequest .. generated/gallery/export_jpg                                                                                             
writing output... [100%] generated/gallery/export_movie .. whatsnew/index                                                                                                          
../CHANGELOG.rst:22: WARNING: py:meth reference target not found: drms.client.Client.Client.export
```

File: CHANGELOG.rst
Comment (nabobalis) [inline CHANGELOG.rst:22]:
```suggestion
- :meth:`drms.client.Client.export()` now accepts a ``process`` keyword argument
```

File: drms/client.py
Comment (nabobalis) [inline drms/client.py:1196]:
```suggestion
            arguments, are validated by the `~drms.client.Client`. In the case of invalid
```

**Review by nabobalis — APPROVED** [review]

**Review by Cadair — APPROVED** [review]

File: CHANGELOG.rst
Comment (nabobalis) [reply CHANGELOG.rst:22]: Yeah. I forgot!

**Pull request merged**
